Fix pyparsing > 3.0 compatibility issue.

parrotpock opened the pull request:
Addresses issue raised in https://github.com/graphite-project/graphite-web/issues/2726 plus one other compatibility issue I came across with a failing piped query test.

I've checked that tests pass but haven't tested this in a live environment. Is it worth adding something to the tox matrix to check multiple pyparsing versions? It's been a while since I've used tox, so I left this part out as I'm not setup to test it properly.

deniszh commented:
Hi @parrotpock ,

Thanks for your contribution!

> Is it worth adding something to the tox matrix to check multiple pyparsing versions? 

As far as I understand tox it's already checking multiple pyprsing versions implicitly. But maybe I'm wrong.

Please also take note, that currently pyparsing version is pinned, so, your change doesn't do much:
https://github.com/graphite-project/graphite-web/blob/master/tox.ini#L39
https://github.com/graphite-project/graphite-web/blob/master/requirements.txt#L46

You need to relax constraits to proper check your changes.

parrotpock commented:
Yeah, I only tested this manually with the updated dependencies. Should I drop the pin in both places or just the tox config for now?

deniszh commented:
Yes, please. Minimal pin has sense, but drop < 3.0.0 part

deniszh commented:
Tests are green, but looks like I was wrong - now it's testing only latest pyparsing, which could be OK, but including also pyparsing2 tests could be good idea too.

ptmcg commented:
As the maintainer of pyparsing, I suggest that at minimum you test against pyparsing 2.4.7 and 3.0.6. 2.4.7 has been the stable release since April, 2020, and 3.0.6 is the most recent released version including all regression and compatibility bug-fixes encountered in 3.0.x.

ptmcg commented on webapp/graphite/render/grammar_unsafe.py:
```diff
@@ -80,7 +80,11 @@
 
 def setRaw(s, loc, toks):
-    toks[0].raw = s[toks[0].start:toks[0].end]
+    try:
```
The code you have in the try-block should be compatible with pyparsing pre-3.0 and post-3.0. It's the code in the except block that only works in pyparsing pre 3.0.

I don't think the code in the try block will *ever* raise `AttributeError`, in either version of pyparsing.

parrotpock replied:
My bad, you're right. I initially had the try/except blocks the other way around, but swapped them to make the post-3.0 compatible change the default and forgot to update the exception handling. I hadn't checked that they both would work post and pre 3.0, and so I fixed this. Thanks for catching this.

ptmcg commented on webapp/graphite/render/evaluator.py:
```diff
@@ -58,7 +58,7 @@ def evaluateTokens(requestContext, tokens, replacements=None, pipedArg=None):
     return evaluateTokens(requestContext, tokens.template, arglist)
 
   if tokens.expression:
-    if tokens.expression.pipedCalls:
```
Good catch on this! It is a subtle change, and was not in the pyparsing release notes (I have added it).

parrotpock replied:
Great. Thanks for the confirmation :)

parrotpock commented:
I added a new target for the suggested pyparsing version and above. Since we removed the upper pin this seems vaguely sensible? 

Thanks for the quick feedback, btw.

deniszh commented:
Tests are green. Thanks a lot, @ptmcg , for suggestions!
And @parrotpock ofc for fixing that.

DanCech reviewed (approved):
Nice!

parrotpock commented:
No problem @deniszh. Let me know if you need anything else from me before you can merge this.

deniszh commented:
I think it's good enough to merge. Thanks!

parrotpock commented:
Thanks @deniszh!

deniszh commented:
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------:|
|❌|1.1.x|Aborted|

To backport manually run: `node scripts/backport --pr 2727`.
For more info read the [Backport documentation](https://github.com/sqren/backport#backport)
